{
  "id" : 346,
  "expression" : "IndexOutOfBoundsException.class",
  "projectName" : "jenkinsci@jenkins",
  "commitID" : "c44fa7f4cf769257844d562353a579a198ea517a",
  "filePath" : "/core/src/test/java/hudson/util/RetrierTest.java",
  "occurrences" : 1,
  "isArithmeticExpression" : 0,
  "isGetTypeMethod" : 0,
  "expressionList" : [ {
    "nodeContext" : "IndexOutOfBoundsException.class",
    "nodeType" : "TypeLiteral",
    "nodePosition" : {
      "charLength" : 31,
      "startLineNumber" : 176,
      "startColumnNumber" : 56,
      "endLineNumber" : 176,
      "endColumnNumber" : 87
    },
    "astNodeNumber" : 3,
    "astHeight" : 3,
    "parentDataList" : [ {
      "locationInParent" : "ChildProperty[org.eclipse.jdt.core.dom.ArrayCreation,initializer]",
      "nodePosition" : {
        "charLength" : 33,
        "startLineNumber" : 176,
        "startColumnNumber" : 55,
        "endLineNumber" : 176,
        "endColumnNumber" : 88
      },
      "nodeContext" : "{IndexOutOfBoundsException.class}",
      "nodeType" : "ArrayInitializer",
      "astNodeNumber" : 4,
      "astHeight" : 4
    }, {
      "locationInParent" : "ChildListProperty[org.eclipse.jdt.core.dom.MethodInvocation,arguments]",
      "nodePosition" : {
        "charLength" : 44,
        "startLineNumber" : 176,
        "startColumnNumber" : 44,
        "endLineNumber" : 176,
        "endColumnNumber" : 88
      },
      "nodeContext" : "new Class[]{IndexOutOfBoundsException.class}",
      "nodeType" : "ArrayCreation",
      "astNodeNumber" : 9,
      "astHeight" : 5
    }, {
      "locationInParent" : "ChildProperty[org.eclipse.jdt.core.dom.MethodInvocation,expression]",
      "nodePosition" : {
        "charLength" : 815,
        "startLineNumber" : 160,
        "startColumnNumber" : 29,
        "endLineNumber" : 176,
        "endColumnNumber" : 89
      },
      "nodeContext" : "new Retrier.Builder<>((Callable<Boolean>)() -> {\n  throw new IndexOutOfBoundsException(\"Exception allowed considered as failure\");\n}\n,(currentAttempt,result) -> result == null ? false : result,ACTION).withAttempts(ATTEMPTS).withDelay(100).withDuringActionExceptions(new Class[]{IndexOutOfBoundsException.class})",
      "nodeType" : "MethodInvocation",
      "astNodeNumber" : 48,
      "astHeight" : 11
    }, {
      "locationInParent" : "ChildProperty[org.eclipse.jdt.core.dom.VariableDeclarationFragment,initializer]",
      "nodePosition" : {
        "charLength" : 882,
        "startLineNumber" : 160,
        "startColumnNumber" : 29,
        "endLineNumber" : 178,
        "endColumnNumber" : 24
      },
      "nodeContext" : "new Retrier.Builder<>((Callable<Boolean>)() -> {\n  throw new IndexOutOfBoundsException(\"Exception allowed considered as failure\");\n}\n,(currentAttempt,result) -> result == null ? false : result,ACTION).withAttempts(ATTEMPTS).withDelay(100).withDuringActionExceptions(new Class[]{IndexOutOfBoundsException.class}).build()",
      "nodeType" : "MethodInvocation",
      "astNodeNumber" : 50,
      "astHeight" : 12
    }, {
      "locationInParent" : "ChildListProperty[org.eclipse.jdt.core.dom.VariableDeclarationStatement,fragments]",
      "nodePosition" : {
        "charLength" : 886,
        "startLineNumber" : 160,
        "startColumnNumber" : 25,
        "endLineNumber" : 178,
        "endColumnNumber" : 24
      },
      "nodeContext" : "r=new Retrier.Builder<>((Callable<Boolean>)() -> {\n  throw new IndexOutOfBoundsException(\"Exception allowed considered as failure\");\n}\n,(currentAttempt,result) -> result == null ? false : result,ACTION).withAttempts(ATTEMPTS).withDelay(100).withDuringActionExceptions(new Class[]{IndexOutOfBoundsException.class}).build()",
      "nodeType" : "VariableDeclarationFragment",
      "astNodeNumber" : 52,
      "astHeight" : 13
    }, {
      "locationInParent" : "ChildListProperty[org.eclipse.jdt.core.dom.Block,statements]",
      "nodePosition" : {
        "charLength" : 904,
        "startLineNumber" : 160,
        "startColumnNumber" : 8,
        "endLineNumber" : 178,
        "endColumnNumber" : 25
      },
      "nodeContext" : "Retrier<Boolean> r=new Retrier.Builder<>((Callable<Boolean>)() -> {\n  throw new IndexOutOfBoundsException(\"Exception allowed considered as failure\");\n}\n,(currentAttempt,result) -> result == null ? false : result,ACTION).withAttempts(ATTEMPTS).withDelay(100).withDuringActionExceptions(new Class[]{IndexOutOfBoundsException.class}).build();\n",
      "nodeType" : "VariableDeclarationStatement",
      "astNodeNumber" : 58,
      "astHeight" : 14
    }, {
      "locationInParent" : "ChildProperty[org.eclipse.jdt.core.dom.MethodDeclaration,body]",
      "nodePosition" : {
        "charLength" : 1794,
        "startLineNumber" : 153,
        "startColumnNumber" : 98,
        "endLineNumber" : 190,
        "endColumnNumber" : 5
      },
      "nodeContext" : "{\n  final int ATTEMPTS=3;\n  final String ACTION=\"print\";\n  RingBufferLogHandler handler=new RingBufferLogHandler(20);\n  Logger.getLogger(Retrier.class.getName()).addHandler(handler);\n  Retrier<Boolean> r=new Retrier.Builder<>((Callable<Boolean>)() -> {\n    throw new IndexOutOfBoundsException(\"Exception allowed considered as failure\");\n  }\n,(currentAttempt,result) -> result == null ? false : result,ACTION).withAttempts(ATTEMPTS).withDelay(100).withDuringActionExceptions(new Class[]{IndexOutOfBoundsException.class}).build();\n  Boolean finalResult=r.start();\n  Assert.assertNull(finalResult);\n  String textNoSuccess=Messages.Retrier_NoSuccess(ACTION,ATTEMPTS);\n  assertTrue(String.format(\"The log should contain '%s'\",textNoSuccess),handler.getView().stream().anyMatch(m -> m.getMessage().contains(textNoSuccess)));\n  String testException=Messages.Retrier_ExceptionFailed(ATTEMPTS,ACTION);\n  assertTrue(String.format(\"The log should contain '%s'\",testException),handler.getView().stream().anyMatch(m -> m.getMessage().startsWith(testException)));\n}\n",
      "nodeType" : "Block",
      "astNodeNumber" : 174,
      "astHeight" : 15
    }, {
      "locationInParent" : "ChildListProperty[org.eclipse.jdt.core.dom.TypeDeclaration,bodyDeclarations]",
      "nodePosition" : {
        "charLength" : 1899,
        "startLineNumber" : 152,
        "startColumnNumber" : 4,
        "endLineNumber" : 190,
        "endColumnNumber" : 5
      },
      "nodeContext" : "@Test public void failedActionWithExceptionAfterThreeAttemptsWithoutListenerTest() throws Exception {\n  final int ATTEMPTS=3;\n  final String ACTION=\"print\";\n  RingBufferLogHandler handler=new RingBufferLogHandler(20);\n  Logger.getLogger(Retrier.class.getName()).addHandler(handler);\n  Retrier<Boolean> r=new Retrier.Builder<>((Callable<Boolean>)() -> {\n    throw new IndexOutOfBoundsException(\"Exception allowed considered as failure\");\n  }\n,(currentAttempt,result) -> result == null ? false : result,ACTION).withAttempts(ATTEMPTS).withDelay(100).withDuringActionExceptions(new Class[]{IndexOutOfBoundsException.class}).build();\n  Boolean finalResult=r.start();\n  Assert.assertNull(finalResult);\n  String textNoSuccess=Messages.Retrier_NoSuccess(ACTION,ATTEMPTS);\n  assertTrue(String.format(\"The log should contain '%s'\",textNoSuccess),handler.getView().stream().anyMatch(m -> m.getMessage().contains(textNoSuccess)));\n  String testException=Messages.Retrier_ExceptionFailed(ATTEMPTS,ACTION);\n  assertTrue(String.format(\"The log should contain '%s'\",testException),handler.getView().stream().anyMatch(m -> m.getMessage().startsWith(testException)));\n}\n",
      "nodeType" : "MethodDeclaration",
      "astNodeNumber" : 182,
      "astHeight" : 16
    } ],
    "currentLineData" : {
      "locationInParent" : "ChildListProperty[org.eclipse.jdt.core.dom.MethodInvocation,arguments]",
      "nodePosition" : {
        "charLength" : 44,
        "startLineNumber" : 176,
        "startColumnNumber" : 44,
        "endLineNumber" : 176,
        "endColumnNumber" : 88
      },
      "nodeContext" : "new Class[]{IndexOutOfBoundsException.class}",
      "nodeType" : "ArrayCreation",
      "astNodeNumber" : 9,
      "astHeight" : 5
    },
    "tokenLength" : 1,
    "type" : "java.lang.Class<java.lang.IndexOutOfBoundsException>"
  } ],
  "positionList" : [ {
    "charLength" : 31,
    "startLineNumber" : 176,
    "startColumnNumber" : 56,
    "endLineNumber" : 176,
    "endColumnNumber" : 87
  } ],
  "layoutRelationDataList" : [ ]
}